File: app/src/main/java/com/geek/infoandroid/android/Level2/les3/ServiceVlaskinChast/MainActivity.java
package com.geek.infoandroid.android.Level2.les3.ServiceVlaskinChast;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    private boolean isConnectBound;
    private BoundService myBoundService;
    private BoundService.MyServiceBinder myServiceBinder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mStartBindServise();

    }

    //создаем сервис коннекшон для обработки событий подключения и отключения от сервиса
    private ServiceConnection boundServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {//вызывается при соединении с сервивсом(в этом методе нам приходит наш сервис)
            myServiceBinder = (BoundService.MyServiceBinder) service;//получаем сервис ,к которому подвязались
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {//при отсоединении от сервиса вызывается этот колбэк
            isConnectBound = false;
            myBoundService = null;
        }
    };


    private void mStartBindServise() {
        //создаем интент чтобы понимать к чему мы хотим привязаться
        Intent intent = new Intent(MainActivity.this,BoundService.class);
        bindService(intent,boundServiceConnection,BIND_AUTO_CREATE);//привязываемся(интент,подключ отключ, флаг автосоздание соединения)
        myServiceBinder.gettt();//получаем данные из сервиса через вложенный класс

    }
}

File: app/src/main/java/com/geek/infoandroid/Kotlin/dagger/MainActivity.kt
package com.example.testingdagger2

import android.app.Application
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import com.geek.infoandroid.R
import javax.inject.Inject

class MainActivity : AppCompatActivity() {

    @Inject//после того как мы в компоненте указали методом куда предоставляем зависимости,зависимости можно получать просто аннотацией @Inject
    lateinit var auto:Auto
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        (application as App).appComponent.inject(this)//после того как заинжектили нашу активити получим все нужные зависимости
//получилось гораздо меньше писанины
        auto.startDep()
        auto.depend1

        (application as App).appComponent.getAuto().startDep()//вызываем класс ауто сразу с зависимостями и вызываем у него метод startDep()
        (application as App).appComponent.getAuto().depend1.start()//получили зависимость Auto у которой вызвали ее зависимость Depend1
    }
}